mllib/src/main/scala/org/apache/spark/mllib/clustering/HierarchicalClustering.scala

```diff
@@ -0,0 +1,627 @@
+      a.getDepth() < b.getDepth() &&
+          breezeNorm(a.center.toBreeze, 2) < breezeNorm(b.center.toBreeze, 2)
+    }
+  }
+
+  /**
+   * Gets the all clusters which are leaves in the cluster tree
+   * @return the Seq of the clusters
+   */
+  def getClusters(): Seq[ClusterTree] = toSeq().filter(_.isLeaf())
+
+  /**
+   * Gets the depth of the cluster in the tree
+   *
+   * @return the depth
+   */
+  def getDepth(): Int = {
+    this.parent match {
+      case None => 0
+      case _ => 1 + this.parent.get.getDepth()
+    }
+  }
+
+  /**
+   * Gets the dendrogram height of the cluster at the cluster tree
+   *
+   * @return the dendrogram height
+   */
+  def getHeight(): Double = {
+    this.children.size match {
+      case 0 => 0.0
+      case _ => this.height.get + this.children.map(_.getHeight()).max
+    }
+  }
+
+  /**
+   * Assigns the closest cluster with a vector
+   * @param metric distance metric
+   * @param v the vector you want to assign to
+   * @return the closest cluster
+   */
+  private[mllib]
+  def assignCluster(metric: Function2[BV[Double], BV[Double], Double])(v: Vector): ClusterTree = {
+    this.children.size match {
+      case 0 => this
+      case size if size > 0 => {
+        val distances = this.children.map(tree => metric(tree.center.toBreeze, v.toBreeze))
+        val minIndex = distances.indexOf(distances.min)
+        this.children(minIndex).assignCluster(metric)(v)
+      }
+    }
+  }
+
+  /**
+   * Assigns the closest cluster index of the clusters with a vector
+   * @param metric distance metric
+   * @param vector the vector you want to assign to
+   * @return the closest cluster index of the all clusters
+   */
+  private[mllib]
+  def assignClusterIndex(metric: Function2[BV[Double], BV[Double], Double])(vector: Vector): Int = {
+    val assignedTree = this.assignCluster(metric)(vector)
+    this.getClusters().indexOf(assignedTree)
+  }
+
+  /**
+   * Gets the number of the clusters in the tree. The clusters are only leaves
+   *
+   * @return the number of the clusters in the tree
+   */
+  def getTreeSize(): Int = this.toSeq().filter(_.isLeaf()).size
+
+  def getVariance(): Option[Double] = this.variance
+
+  def getDataSize(): Option[Long] = this.dataSize
+
+  def getParent(): Option[ClusterTree] = this.parent
+
+  def getChildren(): List[ClusterTree] = this.children
+
+  def isLeaf(): Boolean = (this.children.size == 0)
+
+  /**
+   * The flag that the cluster is splittable
+   *
+   * @return true is splittable
+   */
+  def isSplittable(): Boolean = {
+    this.isLeaf && this.getDataSize != None && this.getDataSize.get >= 2
+  }
+}
+
+/**
+ * Companion object for ClusterTree class
+ */
+object ClusterTree {
+
+  /**
+   * Converts `RDD[Vector]` into a ClusterTree instance
+   *
+   * @param data the data in a cluster
+   * @return a ClusterTree instance
+   */
+  def fromRDD(data: RDD[Vector]): ClusterTree = {
+    val breezeData = data.map(_.toBreeze).cache
```

Comment:
This automatically caches the user's input data. Elsewhere, like in `KMeans`, the behavior is to issue a warning if the input data is not cached, but not to automatically cache it for them. Because you are also automatically caching (and then unpersisting) all split data sets throughout the algorithm, this might be ok, but there should be a note in the doc string explaining that data will be cached by this algorithm.